server/controllers/finance/reports/systemUsage/index.js

```diff
@@ -19,68 +19,117 @@ async function document(req, res, next) {
     const { date } = req.query;
 
     const patientsNbr = `
-      SELECT HOUR(created_at) as 'hour', COUNT(patient.uuid) as 'number'
+      SELECT HOUR(created_at) as 'hour', COUNT(patient.uuid) as 'number',
+        GROUP_CONCAT(DISTINCT user_id ORDER BY user_id) AS users
       FROM patient
       WHERE DATE(created_at) = DATE(?)
       GROUP BY HOUR(created_at)
     `;
 
     const invoicesNbr = `
-      SELECT HOUR(created_at)  as 'hour', COUNT(invoice.uuid) as 'number'
+      SELECT HOUR(created_at) as 'hour', COUNT(invoice.uuid) as 'number',
+        GROUP_CONCAT(DISTINCT user_id ORDER BY user_id) AS users
       FROM invoice
       WHERE DATE(created_at) = DATE(?)
       GROUP BY HOUR(created_at)
     `;
 
     const cashNbr = `
-      SELECT HOUR(created_at)  as 'hour', COUNT(cash.uuid) as 'number'
+      SELECT HOUR(created_at) as 'hour', COUNT(cash.uuid) as 'number',
+        GROUP_CONCAT(DISTINCT user_id ORDER BY user_id) AS users
       FROM cash
       WHERE DATE(created_at) = DATE(?)
       GROUP BY HOUR(created_at)
     `;
 
+    const stockMovementNbr = `
+      SELECT HOUR(created_at) as 'hour', COUNT(DISTINCT document_uuid) as 'number',
+        GROUP_CONCAT(DISTINCT user_id ORDER BY user_id) AS users
+      FROM stock_movement
+      WHERE DATE(created_at) = DATE(?)
+      GROUP BY HOUR(created_at)
+    `;
+
     const [
       patients,
       invoices,
       cashPayments,
+      stockMovements,
     ] = await Promise.all([
       db.exec(patientsNbr, date),
       db.exec(invoicesNbr, date),
       db.exec(cashNbr, date),
+      db.exec(stockMovementNbr, date),
     ]);
 
     const patientMap = _.groupBy(patients, 'hour');
     const cashPaymentMap = _.groupBy(cashPayments, 'hour');
     const invoicesMap = _.groupBy(invoices, 'hour');
+    const stockMovementsMap = _.groupBy(stockMovements, 'hour');
 
     const pivot = [];
 
     for (let i = 0; i < 25; i++) {
       const index = `${i}`;
       const row = {
-        hour : index, patients : 0, invoices : 0, cashPayments : 0,
+        hour : index, patients : 0, invoices : 0, cashPayments : 0, stockMovements : 0, users : 0,
       };
 
+      let aggregateUsers = '';
+
       if (patientMap[index]) {
-        row.patients = patientMap[index][0].number;
+        const { number, users } = patientMap[index][0];
+        row.patients = number;
+        aggregateUsers += users;
       }
 
       if (invoicesMap[index]) {
-        row.invoices = invoicesMap[index][0].number;
+        const { number, users } = invoicesMap[index][0];
+        row.invoices = number;
+        aggregateUsers += users;
       }
 
       if (cashPaymentMap[index]) {
-        row.cashPayments = cashPaymentMap[index][0].number;
+        const { number, users } = cashPaymentMap[index][0];
+        row.cashPayments = number;
+        aggregateUsers += users;
       }
 
+      if (stockMovementsMap[index]) {
+        const { number, users } = cashPaymentMap[index][0];
```

Comment:
I think there is a bug here.  I picked a random day and this line crashed.  You may need to make sure it is not empty before dereferencing it:

      errors #interceptor() TypeError: Cannot read property '0' of undefined
      errors     at document (/home/jmcameron/Bhima/bin/server/controllers/finance/reports/systemUsage/index.js:99:56)
      errors     at processTicksAndRejections (internal/process/task_queues.js:97:5) +0ms

Reply:
Yeah, a pretty serious bug at that, actually.  I'm referencing cash payments instead of stock movements 😞